pkg/beacon/beacon.go

```diff
@@ -175,8 +175,58 @@ func Initialize(
 				onGroupSelected,
 			)
 			if err != nil {
-				logger.Errorf("Tickets submission failed: [%v]", err)
+				logger.Errorf("tickets submission failed: [%v]", err)
 			}
+
+			// We deduplicate `GroupSelectionStarted` events by keeping them
+			// into the `pendingGroupSelections` cache. Duplicated events
+			// may occur due to chain reorgs or because of the subscription
+			// monitoring mechanism which re-fetches past events in case
+			// something bad happens with the websocket connection pointing
+			// to the Ethereum node. However, we can't remove current group
+			// selection from cache just after ticket submission terminates
+			// because a duplicated event coming right after the removal
+			// will trigger a new DKG which will break the current one.
+			// We need to hold the cache entry for the maximum DKG duration
+			// time. We should also do it regardless of the outcome on this
+			// node as other nodes may still perform DKG successfully.
+			// Removing the cache entry after hitting the DKG timeout is
+			// enough and we don't need to monitor for a successful group
+			// registration. A successful group registration means we never
+			// have to trigger group selection again for the group seed we hold
+			// in the cache so we don't have to bother about removing the cache
+			// entry just after the group is registered. In contrary, when
+			// current group selection fails, we need to remove the cache
+			// entry once the timeout is hit to allow the node to retry
+			// the group selection for that seed as soon as possible.
+
+			dkgDurationBlocks :=
+				chainConfig.TicketSubmissionTimeout +
+					gjkr.ProtocolBlocks() +
+					dkgresult.PrePublicationBlocks() +
+					(uint64(chainConfig.GroupSize) * chainConfig.ResultPublicationBlockStep)
+
+			dkgTimeout := event.BlockNumber + dkgDurationBlocks
+
+			if err := blockCounter.WaitForBlockHeight(dkgTimeout); err != nil {
```

Comment:
Can we extract this logic and one for the relay entry to a dedicated deduplicator, just like we have in [`keep-ecdsa`](https://github.com/keep-network/keep-ecdsa/blob/f57fe591a1f14ff95bd22241e5934f48f83c8d0a/pkg/client/event/deduplicator.go)? Having it in a separate struct allows to unit test all cases properly. Also, the time spent on the implementation is not time lost - there is pretty complex logic here calculating dkg duration for the sake of deduplication and I think it deserves to be extracted out of `beacon.go`.

Reply:
But there are already dedicated deduplicators (see `GroupSelectionTrack` and `RelayRequestTrack`). The logic here just determines when to call `Remove` on `GroupSelectionTrack` and belongs to the client code using the deduplicator.

Reply:
I do not think it's the client code, I think it's deduplicator logic to decide when to consider the given event is no longer a duplicate. I'd also love to have this code covered in unit tests, `beacon.go` in future as well. Keeping the code in separate structures would help us with it.

In the deduplicator's notify we'd registered an event in the track struct and scheduled an update to remove it from track when it's no longer needed. Notify completed function in the deduplicator for group selection event is not needed IMO.

Reply:
Ok, I misunderstood. I mixed up things and was sure that when you say "deduplicator" you have `*Track` structures in mind. All clear now, I will extract dedicated deduplicators as we have in `keep-ecdsa`.

Reply:
Ok partially done in: https://github.com/keep-network/keep-core/pull/2574/commits/ea353ddbeff0915b5e9a11c030515c461707ff35. I'll add it for relay entries and add some tests for the deduplicator in further commits.

Reply:
Added for relay entry as well: https://github.com/keep-network/keep-core/pull/2574/commits/fb94fed2acef5178dff02250a02e610347540e02